src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1854 @@
+                    // edits to save and restore reftypes are prepended before the inst
+                    // and after resulting in the following scenario:
+                    //
+                    // --- p0 is overwritten ---
+                    // move from p1 to stack_v0
+                    // 1. use v0 (fixed: p1). Clobbers: [p0]
+                    // move from stack_v0 to p0
+                    // --- p0 is overwritten ---
+                    // 2. use v0 (fixed: p0)
+                    //
+                    // To avoid this scenario, the registers are added to the
+                    // `free_after_curr_inst` instead, to ensure that it isn't used as
+                    // a scratch register.
+                    self.free_after_curr_inst[preg.class()].add(preg);
+                } else {
+                    // Something is still in the clobber.
+                    // After this instruction, it's no longer a clobber.
+                    // Add it back to the LRU.
+                    trace!("Something is still in the clobber {:?}. Adding it back to the LRU directly.", preg);
+                    self.lrus[preg.class()].append_and_poke(preg);
+                }
+            }
+        }
+        trace!("After the allocation:");
+        trace!("freed_def_pregs: {}", self.freed_def_pregs);
+        trace!("free after curr inst: {}", self.free_after_curr_inst);
+        trace!("");
+        let scratch_regs = self.get_scratch_regs(inst)?;
+        self.process_edits(scratch_regs);
+        self.add_freed_regs_to_freelist();
+        self.use_vregs_saved_and_restored_in_curr_inst.clear();
+        self.vregs_first_seen_in_curr_inst.clear();
+        self.vregs_allocd_in_curr_inst.clear();
+        self.reused_inputs_in_curr_inst.clear();
+        self.vregs_in_curr_inst.clear();
+        self.pregs_allocd_in_curr_inst = PRegSet::empty();
+        if trace_enabled!() {
+            self.log_post_inst_processing_state(inst);
+        }
+        Ok(())
+    }
+
+    /// At the beginning of every block, all virtual registers that are
+    /// livein are expected to be in their respective spillslots.
+    /// This function sets the current allocations of livein registers
+    /// to their spillslots and inserts the edits to flow livein values to
+    /// the allocations where they are expected to be before the first
+    /// instruction. 
+    fn reload_at_begin(&mut self, block: Block) {
+        trace!("Reloading live registers at the beginning of block {:?}", block);
+        trace!("Live registers at the beginning of block {:?}: {:?}", block, self.live_vregs);
+        trace!("Block params at block {:?} beginning: {:?}", block, self.func.block_params(block));
+        // We need to check for the registers that are still live.
+        // These registers are livein and they should be stack-allocated.
+        let live_vregs = self.live_vregs.clone();
+        for vreg in live_vregs.iter().cloned() {
```

Comment:
Since this is done at the end of processing for a block, it's faster to just clear end reset all the state at once at the end rather than freeing each allocation individually.

This function could be much simpler:
- remove all blockparams from live_vregs
- for each live vreg move it to its spillslot.

Additionally, you could change the definition of live_vregs to exclude vregs that are in a spillslot. They don't require any special handling here.

Reply:
> Since this is done at the end of processing for a block, it's faster to just clear end reset all the state at once at the end rather than freeing each allocation individually.

I don't think it's that straightforward. `freealloc` does some other things, too, that isn't just clearing stuff.

> This function could be much simpler:

    remove all blockparams from live_vregs
    for each live vreg move it to its spillslot.

Yeah, I'll definitely do that.

> Additionally, you could change the definition of live_vregs to exclude vregs that are in a spillslot. They don't require any special handling here.

And this will require an additional check after every operand allocation to see if it's a stack allocation. So, I'm not sure if the current approach will be faster?